File: src/main/java/com/ject/studytrip/mission/domain/policy/MissionPolicy.java
package com.ject.studytrip.mission.domain.policy;

import com.ject.studytrip.global.exception.CustomException;
import com.ject.studytrip.mission.domain.error.MissionErrorCode;
import com.ject.studytrip.mission.domain.model.Mission;
import java.util.List;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class MissionPolicy {
    public static void validateMissionBelongsToStamp(Long stampId, Mission mission) {
        if (!mission.getStamp().getId().equals(stampId)) {
            throw new CustomException(MissionErrorCode.MISSION_NOT_BELONGS_TO_STAMP);
        }
    }

    public static void validateNotDeleted(Mission mission) {
        if (mission.getDeletedAt() != null) {
            throw new CustomException(MissionErrorCode.MISSION_ALREADY_DELETED);
        }
    }

    public static void validateCompleted(Mission mission) {
        if (mission.isCompleted())
            throw new CustomException(MissionErrorCode.MISSION_ALREADY_COMPLETED);
    }

    public static void validateExistAll(List<Mission> foundMissions, List<Long> requestedIds) {
        boolean isEquals = foundMissions.size() == requestedIds.size();
        if (!isEquals) {
            throw new CustomException(MissionErrorCode.MISSION_NOT_FOUND);
        }
    }

    public static void validateAllCompleted(boolean exists) {
        if (exists) {
            throw new CustomException(MissionErrorCode.ALL_MISSIONS_NOT_COMPLETED);
        }
    }
}

File: src/main/kotlin/com/ject/studytrip/mission/domain/policy/MissionPolicy.kt
package com.ject.studytrip.mission.domain.policy

import com.ject.studytrip.global.exception.CustomException
import com.ject.studytrip.mission.domain.error.MissionErrorCode
import com.ject.studytrip.mission.domain.model.Mission

object MissionPolicy {
    fun validateNotDeleted(mission: Mission) {
        if (mission.isDeleted()) {
            throw CustomException(MissionErrorCode.MISSION_ALREADY_DELETED)
        }
    }

    fun validateNotCompleted(mission: Mission) {
        if (mission.isCompleted()) {
            throw CustomException(MissionErrorCode.MISSION_ALREADY_COMPLETED)
        }
    }

    fun validateNotAllCompleted(exists: Boolean) {
        if (exists) {
            throw CustomException(MissionErrorCode.ALL_MISSIONS_NOT_COMPLETED)
        }
    }

    fun validateMissionBelongsToStamp(
        stampId: Long,
        mission: Mission,
    ) {
        if (mission.stamp.id != stampId) {
            throw CustomException(MissionErrorCode.MISSION_NOT_BELONGS_TO_STAMP)
        }
    }

    fun validateExistAll(
        foundMissions: List<Mission>,
        requestedIds: List<Long>,
    ) {
        if (foundMissions.size != requestedIds.size) {
            throw CustomException(MissionErrorCode.MISSION_NOT_FOUND)
        }
    }
}
